src/ast/SemanticChecksVisitor.java

```diff
@@ -35,32 +42,294 @@ public void visit(Program program) {
+        //check that formals match static type
+        for(int i = 0; i < overriding.formals().size(); i++){
+            if(!sameStaticType(overriding.formals().get(i), overwritten.formals().get(i))){
+                return false;
+            }
+        }
+        //check that return type is covariant
+        if(!isCovariant(overriding.returnType(), overwritten.returnType())) return false;
+
+        return true;
+    }
+
+    /**
+     * given two VariableIntroduction this method checks if they are of the same static type
+     * @param a
+     * @param b
+     * @return
+     */
+    private boolean sameStaticType(VariableIntroduction a, VariableIntroduction b){
+        //not the same ast type - return false
+        if(!a.type().getClass().getName().equals(b.type().getClass().getName())) return false;
+
+        if(a.type().getClass().getName().equals("ast.RefType")){
+            //both reference, check that to the same class
+            return ((RefType)a.type()).id().equals(((RefType)b.type()).id());
+        }
+        //both the same type and not reference - int, boolean or int[]
+        return true;
+    }
+
+    /**
+     * checking for covariance for a given pair of AstType
+     * @param varToCheck
+     * @param toCheckFor
+     * @return varToCheck is an covariant of toCheckFor
+     */
+    private boolean isCovariant(AstType varToCheck, AstType toCheckFor){
+        if(varToCheck.getClass().getName().equals(toCheckFor.getClass().getName())){
+            //both the same ast type - check which one
+            if(!varToCheck.getClass().getName().equals("ast.RefType")){
+                //they are both the same non reference type - either int, boolean or int[]
+                return true;
+            }
+            //now both are identifier, check if there is inheritance
+            return forest.isA(((RefType)varToCheck).id(), ((RefType)toCheckFor).id());
+        }
+        return false;
     }
 
     @Override
     public void visit(MainClass mainClass) {
+        //TODO
+    }
+
+    /**
+     * given an expression this method checks its type and returns it
+     * if it's a reference type this method makes sure to have the correct id set
+     * @param e
+     * @return AstType representing the type of e, null if the expression is not legal
+     */
+    private AstType getExprType(Expr e){
+        if(isIntExpr(e)) return new IntAstType();
+
+        if(isBooleanExpr(e)) return new BoolAstType();
+
+        String dinamicExprName = e.getClass().getName();
+
+        String binaryOps[] = {"ast.AddExpr", "ast.SubtractExpr", "ast.LtExpr", "ast.MultExpr","ast.AndExpr"};
+        if(Arrays.asList(binaryOps).contains(dinamicExprName)){
+            //binary must be either int or bool, so if we're here than it's not a legal expression
+            return null;
+        }
+
+        //int array
+        if(dinamicExprName.equals("ast.NewIntArrayExpr")) return new IntArrayAstType();
```

Comment:
I think this should be moved to be before `binaryOps` and sit with the other primitive type checks